<!DOCTYPE html>
<html lang="en"><head>
<script src="HERN_files/libs/clipboard/clipboard.min.js"></script>
<script src="HERN_files/libs/quarto-html/tabby.min.js"></script>
<script src="HERN_files/libs/quarto-html/popper.min.js"></script>
<script src="HERN_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="HERN_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="HERN_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="HERN_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="HERN_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.3.294">

  <title>Antibody-antigen Docking and Design via Hierarchical Equivariant Refinement Networks (HERN)</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="HERN_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="HERN_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="HERN_files/libs/revealjs/dist/theme/quarto.css" id="theme">
  <link rel="stylesheet" href="HERN_files/libs/quarto-diagram/mermaid.css">
  <link href="HERN_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="HERN_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="HERN_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="HERN_files/libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="HERN_files/libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="HERN_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
  <script src="HERN_files/libs/quarto-diagram/mermaid.min.js"></script>
  <script src="HERN_files/libs/quarto-diagram/mermaid-init.js"></script>
  <link href="HERN_files/libs/quarto-diagram/mermaid.css" rel="stylesheet">
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Antibody-antigen Docking and Design via Hierarchical Equivariant Refinement Networks (HERN)</h1>
  <p class="subtitle">Presentation created by <b>Pejvak Moghimi</b> for InstaDeep</p>

<div class="quarto-title-authors">
</div>

</section>
<section id="background-and-motivations" class="title-slide slide level1 center">
<h1>Background and Motivations</h1>
<div>
<ul>
<li>PhD in machine learning and immunoinformatics
<ul>
<li>Deep noisy long-tailed learning of antibody convergence across AIRRs</li>
</ul></li>
</ul>
</div>
</section>

<section>
<section id="introduction" class="title-slide slide level1 center">
<h1>Introduction</h1>
<ul>
<li class="fragment">Antibodies, Paratopes, Epitopes and docking</li>
<li class="fragment">Graph Neural Networks (GNNs)</li>
<li class="fragment">Message Passing in GNNs</li>
</ul>
</section>
<section id="antibodies-paratopes-epitopes-and-docking" class="slide level2" data-background-color="black" data-background-image="HERN_files/images/figure_antibody_structure_cropped.png" data-background-size="550px" data-background-repeat="repeat" data-background-opacity="0.3">
<h2>Antibodies, Paratopes, Epitopes and docking</h2>
<ul>
<li class="fragment">Antibodies are Y-shaped proteins that bind to antigens.</li>
<li class="fragment">The antibody’s binding-surface is called the paratope.</li>
<li class="fragment">An antigen’s binding-surfaces are called the epitopes.
<ul>
<li class="fragment">An antigen can have multiple epitopes.</li>
</ul></li>
<li class="fragment">Docking is the process of finding the best binding pose of a paratope to an epitope. <img data-src="HERN_files/images/figure_antibody_binding.PNG" class="absolute fragment fade-in-then-out quarto-discovered-preview-image" style="bottom: 10px; right: 1050px; width: 225px; "></li>
</ul>
</section>
<section id="molecules-as-graphs" class="slide level2">
<h2>Molecules as Graphs</h2>
<ul>
<li class="fragment">The nodes in the graph represent the atoms of the molecule.</li>
<li class="fragment">The edges represent the bonds between the atoms.</li>
<li class="fragment">The graph structure captures the spatial information of the molecule.</li>
</ul>
</section>
<section id="section" class="slide level2" data-background-color="white">
<h2></h2>
<p><img data-src="HERN_files/images/caffeine/1680175220705.jpg" class="absolute fragment fade-in-then-out" style="bottom: 50px; right: 30px; width: 900px; "></p>
<p><img data-src="HERN_files/images/caffeine/1680175220679.png" class="absolute fragment" style="bottom: 50px; right: 30px; width: 900px; "></p>
<p><img data-src="HERN_files/images/caffeine/caffeine_cartoon_molecule.png" class="absolute fragment" style="bottom: 50px; right: 30px; width: 900px; "></p>
</section>
<section id="graph-neural-networks-gnns" class="slide level2">
<h2>Graph Neural Networks (GNNs)</h2>
<ul>
<li class="fragment"><p>GNNs are a class of neural networks, based on the idea of message passing, that operate on graph-structured data.</p></li>
<li class="fragment"><p>GNNs have been successfully applied to protein design, molecular docking, drug design, and protein-protein interaction, and other applications.</p></li>
<li class="fragment"><p>GNNs can be used to model complex molecular interactions.</p>
<ul>
<li class="fragment">The atoms (nodes/vertices) of the molecular graph can be represented by a set of features, and the bonds (edges) can also be represented by a set of features.</li>
<li class="fragment">The features of the atoms and bonds can be updated by message passing.</li>
</ul></li>
</ul>
</section>
<section id="equivariant-graph-neural-networks-egnns" class="slide level2">
<h2>Equivariant Graph Neural Networks (EGNNs)</h2>
<ul>
<li class="fragment">Equivariance is a property of a function that allows it to be invariant to a certain transformation.</li>
<li class="fragment">In the context of GNNs, equivariance is a property of the GNN that allows it to be invariant to a certain transformation on the graph.</li>
<li class="fragment">EGNNs are GNNs that are equivariant to graph isomorphism.
<ul>
<li class="fragment">isomorphism is a transformation that preserves the graph structure.</li>
<li class="fragment">I.e. EGNNs are invariant to group structure and the order of nodes in the graph.</li>
</ul></li>
</ul>
</section>
<section id="message-passing-in-gnns" class="slide level2">
<h2>Message Passing in GNNs</h2>
<ul>
<li class="fragment">Message passing is the process of propagating information through a graph.</li>
<li class="fragment">Defines nodes’ and/or edges’ features in terms of the neighbouring nodes’ and/or edges’ features.</li>
<li class="fragment">The message passing is performed in different ways, multiple times in parallel and a hierarchical manner.</li>
<li class="fragment">Nodes’ (Atoms) and edges’ (bonds) features are updated through message passing, resulting in an embedding representation that captures the global and/or local spatial information of the molecule.</li>
<li class="fragment">It can be used in both supervised and unsupervised learning.</li>
</ul>
</section>
<section id="message-passing-in-gnns-1" class="slide level2 scrollable">
<h2>Message Passing in GNNs</h2>
<ul>
<li class="fragment">The message passing process in GNNs is based on the following equation: <span class="math display">\[
  \mathbf{h}_i^{\left(t+1\right)} = \sigma\left(\mathbf{W}_h\mathbf{h}_i^{\left(t\right)} + \sum_{j\in \mathcal{N}(i)} \mathbf{W}_e \mathbf{e}_{ij}^{\left(t\right)}\right)
  \]</span></li>
<li class="fragment">Where <span class="math inline">\(\mathbf{h}_i^{\left(t\right)}\)</span> is the node feature vector at time <span class="math inline">\(t\)</span> for node <span class="math inline">\(i\)</span>, where time is the number of message passing steps.</li>
<li class="fragment"><span class="math inline">\(\mathbf{W}_h\)</span> is the weight matrix for the node features.</li>
<li class="fragment"><span class="math inline">\(\mathbf{e}_{ij}^{\left(t\right)}\)</span> is the edge feature vector at time <span class="math inline">\(t\)</span> between node <span class="math inline">\(i\)</span> and node <span class="math inline">\(j\)</span>.</li>
<li class="fragment"><span class="math inline">\(\mathbf{W}_e\)</span> is the weight matrix for the edge features.</li>
<li class="fragment"><span class="math inline">\(\mathcal{N}(i)\)</span> is the set of neighbors of node <span class="math inline">\(i\)</span>.</li>
<li class="fragment"><span class="math inline">\(\sigma\)</span> is the activation function.</li>
</ul>
</section>
<section id="message-passing-in-gnns-2" class="slide level2">
<h2>Message Passing in GNNs</h2>
<ul>
<li class="fragment">Message passing steps are as follows:
<ul>
<li class="fragment">Reading the node features <span class="math inline">\(\mathbf{h}_i^{\left(t\right)}\)</span>.</li>
<li class="fragment">Reading the edge features <span class="math inline">\(\mathbf{e}_{ij}^{\left(t\right)}\)</span>.</li>
<li class="fragment">Aggregating the node features by an aggregation function <span class="math inline">\(\mathcal{A}\)</span>.</li>
<li class="fragment">Updating the node features by a message function <span class="math inline">\(\mathcal{M}\)</span>.</li>
<li class="fragment">Passing the updated node features to the next node.</li>
<li class="fragment">For each node <span class="math inline">\(i\)</span> in the graph, update its feature vector <span class="math inline">\(h_i\)</span> with the features of its neighbours.</li>
</ul></li>
</ul>
</section>
<section id="problem-formulation" class="slide level2">
<h2>Problem Formulation</h2>
<ul>
<li class="fragment">Given a protein-protein complex and a paratope, find the best docking pose of the paratope to the epitope.</li>
<li class="fragment">The problem is formulated as a graph neural network problem.</li>
<li class="fragment">The input consists of a paratope-epitope represented as a graph.</li>
<li class="fragment">The output consists of the best docking pose of the paratope to the epitope.</li>
</ul>
</section>
<section id="paratope-epitope-complex" class="slide level2">
<h2>Paratope-Epitope Complex</h2>
<ul>
<li class="fragment">The paratope is represented by two graphs:
<ul>
<li class="fragment">A graph to represent the paratope in terms of its residues.
<ul>
<li class="fragment">Each node in the graph only represents <span class="math inline">\(C_{\alpha}\)</span> atom of the residue.</li>
</ul></li>
<li class="fragment">A graph to represent the paratope in terms of its atoms, including the side chains.</li>
</ul></li>
<li class="fragment">Each node in the graph represents an atom.</li>
<li class="fragment">Each edge in the graph represents a bond between two atoms.</li>
<li class="fragment">The nodes in the graph are represented as feature vectors.</li>
<li class="fragment">The features of each node consist of the atomic coordinates.</li>
<li class="fragment">The edges in the graph are represented as feature vectors.</li>
<li class="fragment">The features of each edge consist of the bond type.</li>
</ul>
</section>
<section id="hierarchical-equivariant-refinement-network-hern" class="slide level2">
<h2>Hierarchical Equivariant Refinement Network (HERN)</h2>
<ul>
<li class="fragment">Refinement Networks are neural networks that iteratively refine a set of initial coordinates.</li>
<li class="fragment">The refinement process consists of:
<ul>
<li class="fragment">Predicting the next residue in the sequence.</li>
<li class="fragment">Predicting the next set of coordinates based on the current coordinates and the predicted residue.</li>
<li class="fragment">Computing a loss function based on the predicted and the current coordinates.</li>
</ul></li>
<li class="fragment">The refinement process is repeated until the sequence is fully predicted.</li>
<li class="fragment">The message passing is performed in a hierarchical manner for computational efficieny.</li>
</ul>
</section></section>
<section>
<section id="oversmoothing-and-suspended-animation" class="title-slide slide level1 center">
<h1>Oversmoothing and suspended animation</h1>

<img data-src="HERN_files/animations/caffeine_graph_Gauss-bluring.gif" class="r-stretch"></section>
<section id="docking" class="slide level2">
<h2>Docking</h2>
<div>
<p><img data-src="HERN_files/images/HADDOCKanimation_of_rigid-body_minimization.gif" class="absolute fragment" style="bottom: 110px; right: 130px; width: 300px; "></p>
<p><img data-src="HERN_files/images/haddock_animation_of_refinement_in_explicit_solvent_water.gif" class="absolute" style="top: 170px; left: 30px; width: 400px; height: 400px; "></p>
<p><img data-src="HERN_files/images/haddock_animation_of_semi-flexible_simulated_annealing.gif" class="absolute fragment" style="top: 150px; right: 80px; width: 450px; "></p>
</div>
<div class="footer">
<p>From Bonvin lab: <a href="https://www.bonvinlab.org/education/HADDOCK24/HADDOCK24-antibody-antigen/">HADDOCK2.4 Antibody - Antigen tutorial</a></p>
</div>
</section></section>
<section>
<section id="hern" class="title-slide slide level1 centered-title center">
<h1>HERN</h1>
<style>
/* Center slice title */
.centered-title {
  text-align: center;
}
</style>
</section>
<section id="section-1" class="slide level2">
<h2></h2>
<style>
.center-diagram {
  width: 125%;
  height: 95%;
  display: flex;
  align-items: center;
  justify-content: center;
}
.slide {
  display: flex;
  align-items: center;
  justify-content: center;
}
.node text {
  font-size: 120px;
}
.label text {
  font-size: 30px;
}
</style>
<section class="slide">
<div class="center-diagram">
<div class="cell" data-reveal="true">
<div class="cell-output-display">
<div>
<div>
<pre class="mermaid mermaid-js">classDiagram
  class EGNNEncoder {
    - update_X: bool
    - features_type: str
    - features: ProteinFeatures
    - node_in: int
    - edge_in: int
    - W_v: nn.Linear
    - W_e: nn.Linear
    - layers: nn.ModuleList
    - W_x: nn.Linear
    - U_x: nn.Linear
    - T_x: nn.Sequential
    + __init__(args, node_hdim=0, features_type='backbone', update_X=True)
    - forward(X, V, S, A)

  }
  class HierEGNNEncoder {
    - update_X: bool
    - backbone_CA_only: bool
    - clash_step: int
    - residue_mpn: EGNNEncoder
    - atom_mpn: EGNNEncoder
    - W_x: nn.Linear
    - U_x: nn.Linear
    - T_x: nn.Sequential
    - W_a: nn.Linear
    - U_a: nn.Linear
    - T_a: nn.Sequential
    - embedding: nn.Embedding
    + __init__(args, update_X=True, backbone_CA_only=True)
    + forward(X, V, S, A)
  }
  class MPNNLayer {
    - num_hidden: int
    - num_in: int
    - dropout: nn.Dropout
    - norm: nn.Identity
    - W: nn.Sequential
    + forward(h_V: torch.Tensor, h_E: torch.Tensor, mask_attend: torch.Tensor) -&gt; torch.Tensor
  }
  class PosEmbedding {
    - num_embeddings: int
    + forward(E_idx: torch.Tensor) -&gt; torch.Tensor
  }
  class AAEmbedding {
    - hydropathy: dict
    - volume: dict
    - charge: dict
    - polarity: dict
    - acceptor: dict
    - donor: dict
    - embedding: torch.Tensor
    + to_rbf(D: torch.Tensor, D_min: float, D_max: float, stride: float) -&gt; torch.Tensor
    + transform(aa_vecs: torch.Tensor) -&gt; torch.Tensor
    + dim() -&gt; int
    + forward(x: torch.Tensor, raw: bool=False) -&gt; torch.Tensor
    + soft_forward(x: torch.Tensor) -&gt; torch.Tensor
  }
  class ABModel {
    - k_neighbors: int
    - hidden_size: int
    - embedding: AAEmbedding
    - features: ProteinFeatures
    - W_i: nn.Linear
    - bce_loss: nn.BCEWithLogitsLoss
    - ce_loss: nn.CrossEntropyLoss
    - mse_loss: nn.MSELoss
    - huber_loss: nn.SmoothL1Loss
    + select_target(tgt_X: torch.Tensor, tgt_h: torch.Tensor, tgt_A: torch.Tensor, tgt_pos: list) -&gt; tuple
}
  class PositionalEncodings {
    -num_embeddings: int
    -period_range: list
    +__init__(num_embeddings, period_range=[2,1000])
  }
  class ProteinFeatures {
    - top_k: int
    - num_rbf: int
    - features_type: str
    - direction: str
    + forward(positions: torch.Tensor, h_V: torch.Tensor, h_E: torch.Tensor, A: torch.Tensor, R: torch.Tensor, pos: list) -&gt; tuple
  }
  class RefineDocker {
    + rstep: int
    + target_mpn: EGNNEncoder
    + hierarchical: bool
    - U_i: nn.Linear
    - struct_mpn: Union[EGNNEncoder, HierEGNNEncoder]
    - W_x0: nn.Sequential
    - U_x0: nn.Sequential
    + struct_loss(bind_X, tgt_X, true_V, true_R, true_D, inter_D, true_C): Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]
    + forward(binder, target, surface): ReturnType
  }
  class CondRefineDecoder {
    -hierarchical: bool
    -residue_atom14: Tensor
    -W_s0: Sequential
    -W_x0: Sequential
    -U_x0: Sequential
    -W_s: Linear
    -U_i: Linear
    -target_mpn: EGNNEncoder
    -struct_mpn: EGNNEncoder or HierEGNNEncoder
    -seq_mpn: EGNNEncoder or HierEGNNEncoder
    +struct_loss(): Function
    +forward(): ReturnType
    +generate(): ReturnType
  }
  class AttRefineDecoder {
    - W_x: nn.Linear
    - W_s: nn.Linear
    - struct_mpn: EGNNEncoder
    - seq_mpn: EGNNEncoder
    - W_x0: nn.Sequential
    - U_i: nn.Linear
    - target_mpn: EGNNEncoder
    - W_att: nn.Sequential
    + attention(Q: Tensor, context: Tensor, cmask: Tensor): Tensor
    + struct_loss(X: Tensor, mask: Tensor, true_D: Tensor, true_V: Tensor, true_R: Tensor, true_C: Tensor): Tuple
    + forward(binder: Tuple, target: Tuple, surface: Tuple): ReturnType
    + generate(target: Tuple, surface: Tuple): ReturnType
  }
  class UncondRefineDecoder {
    - W_x: nn.Linear
    - W_s: nn.Linear
    - struct_mpn: EGNNEncoder
    - seq_mpn: EGNNEncoder
    - W_x0: nn.Sequential
    + struct_loss(X: Tensor, mask: Tensor, true_D: Tensor, true_V: Tensor, true_R: Tensor, true_C: Tensor): Tuple
    + forward(binder: Tuple, target: Tuple, surface: Tuple): ReturnType
    + generate(target: Tuple, surface: Tuple): ReturnType
  }
  class SequenceDecoder {
    - no_target
    - W_s
    - seq_rnn
    - U_i
    - target_mpn
    - W_att
    + attention(Q, context, cmask)
    + forward(binder, target, surface)
    + generate(target, surface)
  }
  
  HierEGNNEncoder *-- EGNNEncoder : init
  ABModel *-- AAEmbedding : W_i = nn.Linear(self.embedding.dim(), args.hidden_size)
  ABModel *-- ProteinFeatures : init features
  RefineDocker *-- EGNNEncoder : init
  CondRefineDecoder *-- EGNNEncoder : init
  AttRefineDecoder *-- EGNNEncoder : init
  ABModel --|&gt; RefineDocker : inherits dihedral features
  ABModel --|&gt; CondRefineDecoder : Inherits
  ABModel --|&gt; AttRefineDecoder : Inherits
  RefineDocker *-- HierEGNNEncoder : init
  CondRefineDecoder *-- HierEGNNEncoder : init
  ABModel --|&gt; UncondRefineDecoder : Inherits
  ABModel --|&gt; SequenceDecoder : Inherits
  PositionalEncodings --* ProteinFeatures : init
  PosEmbedding --* AttRefineDecoder : init
  PosEmbedding --* CondRefineDecoder : init
  PosEmbedding --* UncondRefineDecoder : init
  EGNNEncoder &lt;|-- MPNNLayer : init
  EGNNEncoder &lt;|-- ProteinFeatures : init
</pre>
</div>
</div>
</div>
</div>
</div>
</section>
</section></section>
<section>
<section id="equations" class="title-slide slide level1 center">
<h1>Equations</h1>

</section>
<section id="equation-1" class="slide level2">
<h2>equation 1</h2>
<ul>
<li class="fragment">A paratope is a sequence of residuesin the complementarity determining regions (CDRs) of an antibody:
<ul>
<li class="fragment"><span class="math inline">\(\boldsymbol{a}=\boldsymbol{a}_1 \cdots a_n\)</span></li>
</ul></li>
<li class="fragment">An epitope is composed of <span class="math inline">\(m\)</span> residues that are closest to a paratope:
<ul>
<li class="fragment"><span class="math inline">\(\boldsymbol{b}=\boldsymbol{b}_1 \boldsymbol{b}_2 \cdots \boldsymbol{b}_m\)</span></li>
<li class="fragment">It is a subsequence of an antigen:
<ul>
<li class="fragment"><span class="math inline">\(\boldsymbol{c}=\boldsymbol{c}_1 \boldsymbol{c}_2 \cdots \boldsymbol{c}_M\)</span>,</li>
<li class="fragment">where <span class="math inline">\(\boldsymbol{b}_i=\boldsymbol{c}_{e_i}\)</span> and <span class="math inline">\(e_i\)</span> is the index of epitope residue <span class="math inline">\(\boldsymbol{b}_i\)</span> in the antigen.</li>
</ul></li>
</ul></li>
</ul>
</section>
<section id="equation-2" class="slide level2">
<h2>equation 2</h2>
<div style="font-size: 0.85em">
<ul>
<li class="fragment">The epitope 3D structure <span class="math inline">\(\mathcal{G}_b\)</span> is described as a point cloud of atoms:
<ul>
<li class="fragment"><span class="math inline">\(\left\{\boldsymbol{b}_{i, j}\right\}_{1 \leq i \leq m, 1 \leq j \leq n_i}\)</span>
<ul>
<li class="fragment">where <span class="math inline">\(n_i\)</span> is the number of atoms in residue <span class="math inline">\(\boldsymbol{b}_i\)</span>.</li>
</ul></li>
</ul></li>
<li class="fragment">The 3D structure of a paratope and a paratope-epitope binding interface is denoted as <span class="math inline">\(\mathcal{G}_a\)</span> and <span class="math inline">\(\mathcal{G}_{a, b}\)</span>, respectively.
<ul>
<li class="fragment">The first four atoms in any residue correspond to its backbone atoms <span class="math inline">\(\left(\mathrm{N}, \mathrm{C}_\alpha, \mathrm{C}, \mathrm{O}\right)\)</span> and the rest are its side chain atoms.</li>
<li class="fragment">The 3D coordinate of an atom <span class="math inline">\(\boldsymbol{b}_{i, k}\)</span> is denoted as <span class="math inline">\(\boldsymbol{x}\left(\boldsymbol{b}_{i, k}\right) \in \mathbb{R}^3\)</span>.</li>
</ul></li>
</ul>
</div>
</section>
<section id="paratope-docking-and-generation-formulated-as-a-joint-3d-point-cloud-completion-task" class="slide level2 slide-title-60 centered-title">
<h2>Paratope Docking and Generation Formulated as a Joint 3D Point Cloud Completion Task</h2>
<div>
<ul>
<li>Given a training set of antibody-antigen complexes, HERN learns to append an epitope structure <span class="math inline">\(\mathcal{G}_b\)</span> with paratope atoms to form a binding interface <span class="math inline">\(\mathcal{G}_{a, b}\)</span>.</li>
<li><b>put image here</b></li>
</ul>
</div>
</section>
<section id="selecting-epitope-residues" class="slide level2">
<h2>Selecting Epitope Residues</h2>
<ul>
<li class="fragment">The paratope-epitope binding interface is defined as the set of atoms that are within a distance <span class="math inline">\(d\)</span> of each other:
<ul>
<li class="fragment"><span class="math inline">\(\mathcal{G}_{a, b}=\left\{\boldsymbol{a}_{i, j} \mid \boldsymbol{a}_{i, j} \in \mathcal{G}_a, \boldsymbol{b}_{i, k} \in \mathcal{G}_b, \left\|\boldsymbol{x}\left(\boldsymbol{a}_{i, j}\right)-\boldsymbol{x}\left(\boldsymbol{b}_{i, k}\right)\right\| \leq d\right\}\)</span></li>
</ul></li>
<li class="fragment">The paratope-epitope binding interface is a subset of the paratope 3D structure <span class="math inline">\(\mathcal{G}_a\)</span>.</li>
</ul>
</section>
<section id="atom-level-interface-encoder" class="slide level2">
<h2>ATOM-LEVEL INTERFACE ENCODER</h2>
<ul>
<li class="fragment">The atom-level encoder keeps all the atoms in the point cloud <span class="math inline">\(\mathcal{G}_{a, b}\)</span> and constructs a <span class="math inline">\(K\)</span> nearest neighbor graph <span class="math inline">\(\mathcal{G}_{a, b}^A\)</span>. Each node feature is a one-hot encoding of its atom name (e.g., N, <span class="math inline">\(\mathrm{C}_\alpha, \mathrm{C}_\beta, \mathrm{O}\)</span> ).</li>
<li class="fragment">Each edge feature between two atoms <span class="math inline">\(\left(\boldsymbol{a}_{i, k}, \boldsymbol{b}_{j, l}\right)\)</span> is represented as:
<ul>
<li class="fragment"><span class="math display">\[
\boldsymbol{f}\left(\boldsymbol{a}_{i, k}, \boldsymbol{b}_{j, l}\right)=\operatorname{RBF}\left(\left\|\boldsymbol{x}\left(\boldsymbol{a}_{i, k}\right)-\boldsymbol{x}\left(\boldsymbol{b}_{j, l}\right)\right\|\right)
\]</span></li>
</ul></li>
</ul>
</section>
<section id="residue-level-interface-encoder" class="slide level2">
<h2>RESIDUE-LEVEL INTERFACE ENCODER</h2>
<div style="font-size: 0.55em">
<ul>
<li class="fragment">The residue-level encoder only keeps the <span class="math inline">\(\mathrm{C}_\alpha\)</span> atoms and constructs a <span class="math inline">\(K\)</span> nearest neighbor graph <span class="math inline">\(\mathcal{G}_{a, b}^R\)</span> at the residue level.</li>
<li class="fragment">Each residue is represented by a feature vector <span class="math inline">\(\boldsymbol{f}\left(\boldsymbol{a}_i\right)\)</span> including:
<ul>
<li class="fragment">Amino acid features. Each amino acid is represented by six features:
<ul>
<li class="fragment">polarity <span class="math inline">\(f_p \in\{0,1\}\)</span></li>
<li class="fragment">hydropathy <span class="math inline">\(f_h \in[-4.5,4.5]\)</span></li>
<li class="fragment">volume <span class="math inline">\(f_v \in[60.1,227.8]\)</span></li>
<li class="fragment">charge <span class="math inline">\(f_c \in\{-1,0,1\}\)</span></li>
<li class="fragment">whether it is a hydrogen bond donor <span class="math inline">\(f_d \in\{0,1\}\)</span> or acceptor <span class="math inline">\(f_a \in\{0,1\}\)</span>.</li>
</ul></li>
<li class="fragment">For hydropathy and volume features, the radial basis is expanded with interval size 0.1 and 10, respectively</li>
<li class="fragment">→ amino acid feature with 112 dimensions.</li>
</ul></li>
<li class="fragment">It then concatenates the amino acid features with the sum of atom-level <span class="math inline">\(\boldsymbol {h}\left(\boldsymbol{a}_{i, k}\right)\)</span> within that residue as the initial residue representation: <span class="math display">\[
\begin{aligned}
\tilde{\boldsymbol{f}}\left(\boldsymbol{a}_i\right) &amp; =\boldsymbol{f}\left(\boldsymbol{a}_i\right) \oplus \sum_k \boldsymbol{h}\left(\boldsymbol{a}_{i, k}\right) \\
\tilde{\boldsymbol{f}}\left(\boldsymbol{b}_j\right) &amp; =\boldsymbol{f}\left(\boldsymbol{b}_j\right) \oplus \sum_l \boldsymbol{h}\left(\boldsymbol{b}_{j, l}\right)
\end{aligned}
\]</span></li>
</ul>
</div>
</section>
<section id="residue-level-edge-features" class="slide level2">
<h2>Residue-level edge features</h2>
<div style="font-size: 0.65em">
<ul>
<li class="fragment">Next, it builds a local coordinate frame for each residue and compute edge features <span class="math inline">\(\boldsymbol{f}\left(\boldsymbol{a}_i, \boldsymbol{a}_j\right)\)</span> describing the distance, direction, and orientation between nearby residues (Ingraham et al., 2019).</li>
<li class="fragment">The local coordinate frame for each residue <span class="math inline">\(\boldsymbol{a}_i\)</span> is defined as <span class="math inline">\(\boldsymbol{O}_i=\left[\boldsymbol{c}_i, \boldsymbol{n}_i, \boldsymbol{c}_i \times \boldsymbol{n}_i\right]\)</span>, where:
<ul>
<li class="fragment"><span class="math inline">\(\boldsymbol{u}_i=\frac{\boldsymbol{x}_i-\boldsymbol{x}_{i-1}}{\left\|\boldsymbol{x}_i-\boldsymbol{x}_{i-1}\right\|}\)</span></li>
<li class="fragment"><span class="math inline">\(\quad \boldsymbol{c}_i=\frac{\boldsymbol{u}_i-\boldsymbol{u}_{i+1}}{\left\|\boldsymbol{u}_i-\boldsymbol{u}_{i+1}\right\|}\)</span></li>
<li class="fragment"><span class="math inline">\(\quad \boldsymbol{n}_i=\frac{\boldsymbol{u}_i \times \boldsymbol{u}_{i+1}}{\left\|\boldsymbol{u}_i \times \boldsymbol{u}_{i+1}\right\|}\)</span></li>
</ul></li>
</ul>
</div>
</section>
<section id="relative-spatial-encodings" class="slide level2">
<h2>Relative Spatial Encodings</h2>
<div style="font-size: 0.5em">
<ul>
<li class="fragment">Ingraham <em>et al.</em><span class="citation" data-cites="Ingraham2019-bx"><sup><a href="#/refernces" role="doc-biblioref" onclick="">1</a></sup></span> developed invariant and locally informative features</li>
<li class="fragment">by augmenting <span class="math inline">\(\boldsymbol{x}_i\)</span> with ‘orientations’ <span class="math inline">\(\boldsymbol{O}_i\)</span>; defined in terms of the backbone geometry as: <span class="math display">\[
\boldsymbol{O}_i=\left[\begin{array}{lll}
\boldsymbol{b}_i &amp; \boldsymbol{n}_i &amp; \boldsymbol{b}_i \times \boldsymbol{n}_i
\end{array}\right],
\]</span></li>
<li class="fragment">where <span class="math inline">\(\boldsymbol{b}_i\)</span> is the negative bisector of angle between the rays <span class="math inline">\(\left(\boldsymbol{x}_{i-1}-\boldsymbol{x}_i\right)\)</span> and <span class="math inline">\(\left(\boldsymbol{x}_{i+1}-\boldsymbol{x}_i\right)\)</span>, and <span class="math inline">\(\boldsymbol{n}_i\)</span> is a unit vector normal to that plane.</li>
<li class="fragment">Formally: <span class="math display">\[
\boldsymbol{u}_i=\frac{\boldsymbol{x}_i-\boldsymbol{x}_{i-1}}{\left\|\boldsymbol{x}_i-\boldsymbol{x}_{i-1}\right\|}, \quad \boldsymbol{b}_i=\frac{\boldsymbol{u}_i-\boldsymbol{u}_{i+1}}{\left\|\boldsymbol{u}_i-\boldsymbol{u}_{i+1}\right\|}, \quad \boldsymbol{n}_i=\frac{\boldsymbol{u}_i \times \boldsymbol{u}_{i+1}}{\left\|\boldsymbol{u}_i \times \boldsymbol{u}_{i+1}\right\|}
\]</span></li>
<li class="fragment">Finally, they derived the spatial edge features <span class="math inline">\(e_{i j}^{(s)}\)</span> from the rigid body transformation that relates reference frame <span class="math inline">\(\left(\boldsymbol{x}_i, \boldsymbol{O}_i\right)\)</span> to reference frame <span class="math inline">\(\left(\boldsymbol{x}_j, \boldsymbol{O}_j\right)\)</span>.</li>
<li class="fragment">Then, decomposed it into features for distance, direction, and orientation as <span class="math display">\[
\boldsymbol{e}_{i j}^{(s)}=\left(\mathbf{RBF}\left(\left\|\boldsymbol{x}_j-\boldsymbol{x}_i\right\|\right), \quad \boldsymbol{O}_i^T \frac{\boldsymbol{x}_j-\boldsymbol{x}_i}{\left\|\boldsymbol{x}_j-\boldsymbol{x}_i\right\|}, \quad \mathbf{q}\left(\boldsymbol{O}_i^T \boldsymbol{O}_j\right)\right)
\]</span></li>
<li class="fragment">Here the first vector is a distance encoding <span class="math inline">\(\mathbf{RBF}(\cdot)\)</span> lifted into a radial basis, the second vector is a direction encoding that corresponds to the relative direction of <span class="math inline">\(\boldsymbol{x}_j\)</span> in the reference frame of <span class="math inline">\(\left(\boldsymbol{x}_i, \boldsymbol{O}_i\right)\)</span>, and the third vector is an orientation encoding <span class="math inline">\(\mathbf{q}(\cdot)\)</span> of the quaternion representation of the spatial rotation matrix <span class="math inline">\(\boldsymbol{O}_i^T \boldsymbol{O}_j\)</span>.</li>
</ul>
</div>
<aside class="notes">
<ul>
<li>by first augmenting the points <span class="math inline">\(\boldsymbol{x}_i\)</span> with ‘orientations’ <span class="math inline">\(\boldsymbol{O}_i\)</span> that define a local coordinate system at each point; defined in terms of the backbone geometry as</li>
<li></li>
<li><span class="math inline">\(\boldsymbol{u}_i=\frac{\boldsymbol{x}_i-\boldsymbol{x}_{i-1}}{\left\|\boldsymbol{x}_i-\boldsymbol{x}_{i-1}\right\|}\)</span> is the direction.</li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
<div class="slide-footer">
[1] Ingraham <em>et al.</em>, Adv. Neural Inf. Process. Syst., 2019, <a href="https://papers.nips.cc/paper_files/paper/2019/hash/f3a4ff4839c56a5f460c88cce3666a2b-Abstract.html" target="_blank">DOI</a>
</div>
</section>
<section id="quaternions" class="slide level2 centered-title scrollable">
<h2>Quaternions</h2>
<div style="font-size: 0.75em">
<ul>
<li class="fragment">Quaternions represent 3D rotations as four-element vectors that can be efficiently and reasonably compared by inner products.</li>
<li class="fragment">utilises complex numbers for geometric transformations/rotations in 3D space</li>
<li class="fragment">Better than Euler angles
<ul>
<li class="fragment">Does not suffer from Gimble Lock
<ul>
<li class="fragment">Gimble Lock is the case when the Euler angles are not unique and the rotation is not well defined</li>
</ul></li>
</ul></li>
</ul>
</div>
</section>
<section id="section-2" class="slide level2">
<h2></h2>
<p><video data-src="HERN_files/animations/Quaternions.mp4" controls=""><a href="HERN_files/animations/Quaternions.mp4">Video</a></video></p>
<div class="footer">
<p><a href="https://eater.net/quaternions"><b>3blue1brown</b></a></p>
</div>
</section>
<section id="edge-feature-computation" class="slide level2">
<h2>edge feature computation</h2>
<div style="font-size: 0.65em">
<ul>
<li class="fragment">Based on the local frame, we compute the following edge features
<ul>
<li class="fragment"><span class="math display">\[
\boldsymbol{f}\left(\boldsymbol{a}_i, \boldsymbol{a}_j\right)=\left(E_{\mathrm{pos}}(i-j), \quad \operatorname{RBF}\left(\left\|\boldsymbol{x}_{i, 1}-\boldsymbol{x}_{j, 1}\right\|\right), \quad \boldsymbol{O}_i^{\top} \frac{\boldsymbol{x}_{j, 1}-\boldsymbol{x}_{i, 1}}{\left\|\boldsymbol{x}_{i, 1}-\boldsymbol{x}_{j, 1}\right\|}, \quad \boldsymbol{q}\left(\boldsymbol{O}_i^{\top} \boldsymbol{O}_j\right)\right)
\]</span></li>
</ul></li>
</ul>
</div>
<aside class="notes">
<ul>
<li><u><b>Relative positional encodings</b></u>:
<ul>
<li>As in the original Transformer, we also represent distances between residues in the sequence (rather than space) with positional embeddings <span class="math inline">\(\boldsymbol{e}_{i j}^{(p)}\)</span>.</li>
<li>Positioning of each neighbor <span class="math inline">\(j\)</span> is relative to the node under consideration <span class="math inline">\(i\)</span></li>
<li>As result, we get a position embedding as a sinusoidal function of the gap <span class="math inline">\(i-j\)</span>.</li>
</ul></li>
</ul>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
</section>
<section id="message-passing" class="slide level2">
<h2>Message Passing</h2>
<div style="font-size: 0.65em">
<ul>
<li class="fragment"><span class="math inline">\(\mathcal{G}_{a, b}^A\)</span> is encoded by a MPN which learns a vector representation <span class="math inline">\(\boldsymbol{h}\left(\boldsymbol{a}_{i, k}\right), \boldsymbol{h}\left(\boldsymbol{b}_{j, l}\right)\)</span> for each paratope atom <span class="math inline">\(\boldsymbol{a}_{i, k}\)</span> and epitope atom <span class="math inline">\(\boldsymbol{b}_{j, l}\)</span>.</li>
<li class="fragment"><span class="math display">\[
  \boldsymbol{h}_{l+1}\left(\boldsymbol{a}_i\right)=\boldsymbol{h}_l\left(\boldsymbol{a}_i\right)+\sum_{j \in \mathcal{N}_i} \operatorname{FFN}\left(\boldsymbol{h}_l\left(\boldsymbol{a}_i\right), \boldsymbol{h}_l\left(\boldsymbol{a}_j\right), \boldsymbol{f}\left(\boldsymbol{a}_j\right), \boldsymbol{f}\left(\boldsymbol{a}_i, \boldsymbol{a}_j\right)\right) \quad(0 \leq l&lt;L)
  \]</span></li>
<li class="fragment"><span class="math inline">\(\boldsymbol{h}_l\left(\boldsymbol{a}_i\right)\)</span> is the vector representation of the <span class="math inline">\(i\)</span>-th atom <span class="math inline">\(\boldsymbol{a}_i\)</span> after <span class="math inline">\(l\)</span> message passing steps.</li>
</ul>
</div>
</section>
<section id="refernces" class="slide level2 smaller scrollable">
<h2>Refernces</h2>
<div class="footer footer-default">
<p><a href="https://github.com/Peji-moghimi/sciprez" class="uri">https://github.com/Peji-moghimi/sciprez</a></p>
</div>
<div id="refs" class="references csl-bib-body" data-line-spacing="2" role="list">
<div id="ref-Ingraham2019-bx" class="csl-entry" role="listitem">
<div class="csl-left-margin">1. </div><div class="csl-right-inline">Ingraham, J., Garg, V., Barzilay, R. &amp; Jaakkola, T. Generative models for graph-based protein design. <em>Adv. Neural Inf. Process. Syst.</em> <strong>32</strong>, (2019).</div>
</div>
<div id="ref-Eastman2017-gj" class="csl-entry" role="listitem">
<div class="csl-left-margin">2. </div><div class="csl-right-inline">Eastman, P. <em>et al.</em> <span>OpenMM</span> 7: Rapid development of high performance algorithms for molecular dynamics. <em>PLoS Comput. Biol.</em> <strong>13</strong>, e1005659 (2017).</div>
</div>
<div id="ref-Akbar2022-uy" class="csl-entry" role="listitem">
<div class="csl-left-margin">3. </div><div class="csl-right-inline">Akbar, R. <em>et al.</em> Progress and challenges for the machine learning-based design of fit-for-purpose monoclonal antibodies. <em>MAbs</em> <strong>14</strong>, 2008790 (2022).</div>
</div>
<div id="ref-Vogt2014-ij" class="csl-entry" role="listitem">
<div class="csl-left-margin">4. </div><div class="csl-right-inline">Vogt, A. D., Pozzi, N., Chen, Z. &amp; Di Cera, E. Essential role of conformational selection in ligand binding. <em>Biophys. Chem.</em> <strong>186</strong>, 13–21 (2014).</div>
</div>
<div id="ref-Sela-Culang2013-ok" class="csl-entry" role="listitem">
<div class="csl-left-margin">5. </div><div class="csl-right-inline">Sela-Culang, I., Kunik, V. &amp; Ofran, Y. The structural basis of antibody-antigen recognition. <em>Front. Immunol.</em> <strong>4</strong>, 302 (2013).</div>
</div>
<div id="ref-Tsuchiya2016-wc" class="csl-entry" role="listitem">
<div class="csl-left-margin">6. </div><div class="csl-right-inline">Tsuchiya, Y. &amp; Mizuguchi, K. The diversity of <span>H3</span> loops determines the antigen-binding tendencies of antibody <span>CDR</span> loops. <em>Protein Sci.</em> <strong>25</strong>, 815–825 (2016).</div>
</div>
<div id="ref-Jin2021-pj" class="csl-entry" role="listitem">
<div class="csl-left-margin">7. </div><div class="csl-right-inline">Jin, W., Wohlwend, J., Barzilay, R. &amp; Jaakkola, T. <a href="https://arxiv.org/abs/2110.04624">Iterative refinement graph neural network for antibody sequence-structure co-design</a>. (2021).</div>
</div>
<div id="ref-Kim2023-hu" class="csl-entry" role="listitem">
<div class="csl-left-margin">8. </div><div class="csl-right-inline">Kim, J., McFee, M., Fang, Q., Abdin, O. &amp; Kim, P. M. Computational and artificial intelligence-based methods for antibody development. <em>Trends Pharmacol. Sci.</em> <strong>44</strong>, 175–189 (2023).</div>
</div>
<div id="ref-Hummer2022-ut" class="csl-entry" role="listitem">
<div class="csl-left-margin">9. </div><div class="csl-right-inline">Hummer, A. M., Abanades, B. &amp; Deane, C. M. Advances in computational structure-based antibody design. <em>Curr. Opin. Struct. Biol.</em> <strong>74</strong>, 102379 (2022).</div>
</div>
<div id="ref-Khetan2022-xu" class="csl-entry" role="listitem">
<div class="csl-left-margin">10. </div><div class="csl-right-inline">Khetan, R. <em>et al.</em> Current advances in biopharmaceutical informatics: Guidelines, impact and challenges in the computational developability assessment of antibody therapeutics. <em>MAbs</em> <strong>14</strong>, 2020082 (2022).</div>
</div>
<div id="ref-Abanades2022-ad" class="csl-entry" role="listitem">
<div class="csl-left-margin">11. </div><div class="csl-right-inline">Abanades, B., Wong, W. K., Boyles, F., Georges, G., <em>et al.</em> <span>ImmuneBuilder</span>: <span>Deep-Learning</span> models for predicting the structures of immune proteins. <em>bioRxiv</em> (2022).</div>
</div>
<div id="ref-Chinery2023-gj" class="csl-entry" role="listitem">
<div class="csl-left-margin">12. </div><div class="csl-right-inline">Chinery, L., Wahome, N., Moal, I. &amp; Deane, C. M. Paragraph—antibody paratope prediction using graph neural networks with minimal feature vectors. <em>Bioinformatics</em> (2023).</div>
</div>
<div id="ref-Khan2023-tq" class="csl-entry" role="listitem">
<div class="csl-left-margin">13. </div><div class="csl-right-inline">Khan, A. <em>et al.</em> Toward real-world automated antibody design with combinatorial bayesian optimization. <em>Cell Rep Methods</em> <strong>3</strong>, 100374 (2023).</div>
</div>
<div id="ref-Zhou2020-op" class="csl-entry" role="listitem">
<div class="csl-left-margin">14. </div><div class="csl-right-inline">Zhou, J. <em>et al.</em> Graph neural networks: A review of methods and applications. <em>AI Open</em> <strong>1</strong>, 57–81 (2020).</div>
</div>
<div id="ref-Yuan2020-cp" class="csl-entry" role="listitem">
<div class="csl-left-margin">15. </div><div class="csl-right-inline">Yuan, M. <em>et al.</em> A highly conserved cryptic epitope in the receptor binding domains of <span>SARS-CoV-2</span> and <span>SARS-CoV</span>. <em>Science</em> <strong>368</strong>, 630–633 (2020).</div>
</div>
<div id="ref-Fernandez-Quintero2023-on" class="csl-entry" role="listitem">
<div class="csl-left-margin">16. </div><div class="csl-right-inline">Fernández-Quintero, M. L. <em>et al.</em> Challenges in antibody structure prediction. <em>MAbs</em> <strong>15</strong>, 2175319 (2023).</div>
</div>
<div id="ref-Del_Vecchio2021-jx" class="csl-entry" role="listitem">
<div class="csl-left-margin">17. </div><div class="csl-right-inline">Del Vecchio, A., Deac, A., Liò, P. &amp; Veličković, P. <a href="https://arxiv.org/abs/2106.00757">Neural message passing for joint paratope-epitope prediction</a>. (2021).</div>
</div>
<div id="ref-Prince2023-si" class="csl-entry" role="listitem">
<div class="csl-left-margin">18. </div><div class="csl-right-inline">Prince, S. J. D. <em>Understanding deep learning</em>. vol. 1 (2023).</div>
</div>
<div id="ref-Jin2022-wg" class="csl-entry" role="listitem">
<div class="csl-left-margin">19. </div><div class="csl-right-inline">Jin, W., Barzilay, R. &amp; Jaakkola, T. <a href="https://arxiv.org/abs/2207.06616"><span>Antibody-Antigen</span> docking and design via hierarchical equivariant refinement</a>. (2022).</div>
</div>
</div>
</section></section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="HERN_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="HERN_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="HERN_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="HERN_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="HERN_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="HERN_files/libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="HERN_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="HERN_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="HERN_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="HERN_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="HERN_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: true,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>